main.cpp

```diff
@@ -163,98 +171,414 @@ good to go!
+    return divide(*rhs.valuePtr);
+}
+
+DoubleType& DoubleType::add(double rhs)
+{
+    *valuePtr += rhs;
+    return *this;
+}
+
+DoubleType& DoubleType::subtract(double rhs)
+{
+    *valuePtr -= rhs;
+    return *this;
+}
+
+ DoubleType& DoubleType::multiply(double rhs)
+{
+    *valuePtr *= rhs;
+    return *this;
+}
+
+DoubleType& DoubleType::divide(double rhs)
+{
+    *valuePtr /= rhs;
+    return *this;
+}
+
+DoubleType& DoubleType::add(const FloatType& rhs)
+{
+    return add(*rhs.valuePtr);
+}
+
+DoubleType& DoubleType::subtract(const FloatType& rhs)
+{
+    return subtract(*rhs.valuePtr);
+}
+
+ DoubleType& DoubleType::multiply(const FloatType& rhs)
 {
-    return lhs + rhs;
+    return multiply(*rhs.valuePtr);
 }
 
-float FloatType::subtract(float lhs, float rhs)
+DoubleType& DoubleType::divide(const FloatType& rhs)
 {
-    return lhs + rhs;
+    return divide(*rhs.valuePtr);
 }
 
-float FloatType::multiply(float lhs, float rhs)
+DoubleType& DoubleType::add(const DoubleType& rhs)
 {
-    return lhs * rhs;
+    return add(*rhs.valuePtr);
 }
 
-float FloatType::divide(float lhs, float rhs)
+DoubleType& DoubleType::subtract(const DoubleType& rhs)
 {
-    return lhs / rhs;
+    return subtract(*rhs.valuePtr);
 }
 
-double DoubleType::add(double lhs, double rhs)
+ DoubleType& DoubleType::multiply(const DoubleType& rhs)
 {
-    return lhs + rhs;
+    return multiply(*rhs.valuePtr);
 }
 
-double DoubleType::subtract(double lhs, double rhs)
+DoubleType& DoubleType::divide(const DoubleType& rhs)
 {
-    return lhs - rhs;
+    return divide(*rhs.valuePtr);
 }
 
-double DoubleType::multiply(double lhs, double rhs)
+DoubleType& DoubleType::add(const IntType& rhs)
 {
-    return lhs * rhs;
+    return add(*rhs.valuePtr);
 }
 
-double DoubleType::divide(double lhs, double rhs)
+DoubleType& DoubleType::subtract(const IntType& rhs)
 {
-    return lhs / rhs;
+    return subtract(*rhs.valuePtr);
 }
 
-int IntType::add(int lhs, int rhs)
+ DoubleType& DoubleType::multiply(const IntType& rhs)
 {
-    return lhs + rhs;
+    return multiply(*rhs.valuePtr);
 }
 
-int IntType::subtract(int lhs, int rhs)
+DoubleType& DoubleType::divide(const IntType& rhs)
 {
-    return lhs - rhs;
+    return divide(*rhs.valuePtr);
 }
 
-int IntType::multiply(int lhs, int rhs)
+IntType& IntType::add(int rhs)
 {
-    return lhs * rhs;
+    *valuePtr += rhs;
+    return *this;
 }
 
-int IntType::divide(int lhs, int rhs)
+IntType& IntType::subtract(int rhs)
 {
-    if (rhs == 0)
+    *valuePtr -= rhs;
+    return *this;
+}
+
+ IntType& IntType::multiply(int rhs)
+{
+    *valuePtr *= rhs;
+    return *this;
+}
+
+IntType& IntType::divide(int rhs)
+{
+    if(rhs != 0)
     {
-        std::cout << "dividing by 0 is not allowed in this type ! ";
-        return 0;
+        *valuePtr /= rhs;
+        return *this;
     }
     else
     {
-        return lhs / rhs;
-    } 
+        std::cout << "dividing by 0 is not allowed in this type ! " << std::endl;
+        return *this;
+    }
+}
```

Comment:
D.R.Y. 
You can move the `return *this;` outside of the `if/else` blocks.
- [ ] revise